### PR TITLE
Adding h, p and u to basic aer

### PR DESCRIPTION
**Pull request by jaygambetta**
### Summary

While statevector in quantum-info replaces most of the need of basic aer it does not replace QasmSimulator nor should it. I personally think we should remove the unitary and statevector from basicaer but that is a longer term issue. I know that we can use the transpile to rewrite into u1, u2, and u3 these simulations but I am giving some beginner demonstrators and not having h, p and u seems a little strange to require an extra line for using these (transpile). Furthermore since we want to depreciate u1,u2, and u3 it would be good to include all basis [h, p], u and [sx, rz] in basic aer. 

### Details and comments

@blakejohnson I can here you while I'm in the plane see this is why we need simulators :-)

### CODE REVIEW

**Review by 1ucian0 — COMMENTED** [review]
I added a release note.

In general, LGTM. I'm not sure about `u` and its redundancy with `u3` (should it be removed from the basisgates?) and `U`.

**Comment by jaygambetta** [comment]
I would love to remove u3 but I did not want to force it now and break people code that have used u3

**Comment by jakelishman** [comment]
We removed the `QuantumCircuit.u3` method (and the other Euler-angle gates), so it's already pretty hidden.  Actually removing the `U3Gate` class would break a huge amount of extant code, since it's a base gate in both the OQ2 and OQ3 standard libraries (in as much as OQ2 _has_ a standard library), and it doesn't cost us anything to maintain the simple gate definition, which makes it much easier for users to on-ramp to Qiskit.  Since we're committed to supporting OpenQASM with first-class support in Qiskit, it would be a bit strange for us to remove the native representation of one of their standard gates.

We could reduce our internal dependence on it a little more by rewriting more of the standard equivalence libraries to avoid it, but even there, it's only really used as an intermediary in the graph-search algorithm, so it never gets output unless a user's backend explicitly supports it.

**Comment by jaygambetta** [comment]
Jake I agree and It’s why I left it.  My comment was more if we wanted to have a min set in basic air I would put [h p], [sx rz] and [u] and [U] then we could just say for anything more complicated use transpile. Today the h p breaks so why I added this.  

I am also totally ok leaving u1 u2 and u3 as it does not require much work to maintain it and the performance of a longer list to search I doubt is that much.

File: releasenotes/notes/h_basic_aer-3fc5e6776f0de9c1.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+features:
```
Comment (1ucian0) [inline releasenotes/notes/h_basic_aer-3fc5e6776f0de9c1.yaml]:
```suggestion
upgrade:
```

**Comment by jakelishman** [comment]
Jay: I'm sorry - I got this discussion confused with other historical discussions about removing `u1`/`u2`/`u3` in general from Qiskit.

For "which basis gates should be in `BasicAer`?", I guess to me it depends more on what the purpose of it really is.  When I joined, I'd understood it as a sort of "test backend" (i.e. #4443), in which case imo it doesn't hugely matter what the basis is because somebody should always call `transpile` (and it's a noiseless simulator, so gate-specific error/coupling concerns don't come in).  At the time of #4443 and when it was discussed later, this seemed to be the way everyone wanted.

If it is going to be kept and expanded, then it may be more appropriate to do something like #7670 to let it handle almost all Qiskit instructions, not just a restricted set of basis gates?

**Comment by jaygambetta** [comment]
Jake - my understand and I agree with @blakejohnson @levbishop and @chriseclectic is we really need three primitives 

1. Estimator 
2. Sampler
3. run a single Qasm3 and return output variables. 

I believe the plan is the 3 is how backend.run should evolve and what that means is we will need at least basic aer for a QASM simulator. I have no problem with it using statevetor inside but I would like a simple backend.run to run code that supports simple gates sets h, p and rz and sx cx 

Thanks

**Comment by jaygambetta** [comment]
Ie this code should work 

```
from qiskit import QuantumCircuit
from qiskit.providers.basicaer.qasm_simulator import QasmSimulatorPy
from qiskit.compiler import transpile

backend = QasmSimulatorPy()

circuit = QuantumCircuit(1,1)
circuit.h(0)
circuit.measure([0], [0])

result = backend.run(circuit).result()
counts = result.get_counts()
print(counts)
```

**Review by 1ucian0 — APPROVED** [review]
Trying to leave deeper discussions out for avoiding blocking, as we agree that "which basis gates should be in BasicAer" can include gates we currently have in https://github.com/Qiskit/qiskit/tree/main/qiskit/circuit/library/standard_gates, let's merge this PR

**Pull request merged**